src/Symfony/Component/HttpClient/Retry/HttpStatusCodeDecider.php

```diff
@@ -11,25 +11,36 @@
 
 namespace Symfony\Component\HttpClient\Retry;
 
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+
 /**
  * Decides to retry the request when HTTP status codes belong to the given list of codes.
  *
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
 final class HttpStatusCodeDecider implements RetryDeciderInterface
 {
+    private const IDEMPOTENT_METHODS = ['GET', 'HEAD', 'PUT', 'DELETE', 'OPTIONS', 'TRACE', 'CONNECT'];
+
     private $statusCodes;
+    private $idempotentStatusCodes;
 
     /**
-     * @param array $statusCodes List of HTTP status codes that trigger a retry
+     * @param array $statusCodes           List of HTTP status codes that trigger a retry regardless the method
+     * @param array $idempotentStatusCodes List of HTTP status codes that trigger a retry on idempotent methods
      */
-    public function __construct(array $statusCodes = [423, 425, 429, 500, 502, 503, 504, 507, 510])
+    public function __construct(array $statusCodes = [423, 425, 429, 503], array $idempotentStatusCodes = [500, 502, 504, 507, 510])
```

Comment:
502 & 504 can be retried on POST also to me, why not list them?

array $statusCodes = [423 => '*', 425 => '*', 429 => '*', 503 => '*',
    507 => self::IDEMPOTENT_METHODS (const to be made public), etc
]

we could also put the default value in a public const

Reply:
I believe 504 should not be retried on POST: A timeout from the gateway does not mean the backend finished the response nor it failed. It could successfully finish the response but exceed the gateway TTL.